glib: add Error::into_raw()

pbor opened the pull request:
Fixes https://github.com/gtk-rs/gtk-rs/issues/55

GuillaumeGomez commented on glib/src/error.rs:
```diff
@@ -82,6 +83,12 @@ impl Error {
         }
     }
+
+    pub fn into_raw(mut self) -> *mut ffi::GError {
```
I think documentation is needed here, considering that it's calling "mem::forget" on `self`. Also: shouldn't this be an `unsafe` function?

sdroege replied:
No, doesn't have to be `unsafe`

sdroege commented on glib/src/error.rs:
```diff
@@ -82,6 +83,12 @@ impl Error {
+
+    pub fn into_raw(mut self) -> *mut ffi::GError {
+        let e = self.to_glib_none_mut().0;
+        mem::forget(self);
```
Take a look at `Value::into_raw`. It's nicer to use `mem::ManuallyDrop` :)

sdroege commented on glib/src/error.rs:
```diff
@@ -82,6 +83,12 @@ impl Error {
+        let e = self.to_glib_none_mut().0;
+        mem::forget(self);
+        e
+    }
```
These can all be updated to use the new function. Maybe I missed some elsewhere

```
gio/src/subclass/io_stream.rs
176:            *err = e.to_glib_none_mut().0;

gio/src/subclass/seekable.rs
213:            *err = e.to_glib_none_mut().0;
248:            *err = e.to_glib_none_mut().0;

gio/src/subclass/output_stream.rs
234:            *err = e.to_glib_none_mut().0;
258:            *err = e.to_glib_none_mut().0;
282:            *err = e.to_glib_none_mut().0;
314:            *err = e.to_glib_none_mut().0;

gio/src/subclass/input_stream.rs
193:            *err = e.to_glib_none_mut().0;
217:            *err = e.to_glib_none_mut().0;
251:            *err = e.to_glib_none_mut().0;
```

bilelmoussaoui replied:
We also have some in gtk4-rs, it would be very nice if you could update those as well otherwise an issue at least would be enough

pbor commented:
By the way, value has `#[doc(hidden)]`... any reason for that? should I go ahead and document both?

sdroege commented:
> By the way, value has `#[doc(hidden)]`... any reason for that? should I go ahead and document both?

Sounds like a good idea

pbor commented:
Updated, but I guess I am still missing something: I added a commit that calls `g_error_free` on the raw pointer, as I would expect a C caller to do, and it crashes with a double free

sdroege commented on glib/src/value.rs:
```diff
@@ -1224,6 +1224,15 @@ mod tests {
+        unsafe {
+            let v = 123.to_value().into_raw();
+            assert_eq!(gobject_ffi::g_type_check_value(&v), ffi::GTRUE);
+            assert_eq!(gobject_ffi::g_value_get_int(&v), 123);
```
Should probably for completeness also call `gobject_ffi::g_value_unset(&mut v)` here

sdroege commented on glib/src/error.rs:
```diff
@@ -206,6 +213,25 @@ impl error::Error for BoolError {}
+            );
+
+            // THIS CAUSES double free
+            ffi::g_error_free(e);
```
Code looks correct, unclear why this fails. Will debug now

sdroege commented on glib/src/error.rs:
```diff
@@ -206,6 +213,25 @@ impl error::Error for BoolError {}
+            assert_eq!((*e).domain, ffi::g_file_error_quark());
+            assert_eq!((*e).code, ffi::G_FILE_ERROR_FAILED);
+            assert_eq!(
+                CString::from_raw((*e).message),
+                CString::new("Failed").unwrap()
```
```suggestion
                CStr::from_ptr((*e).message),
                CString::new("Failed").unwrap().as_c_str()
```

sdroege replied:
You were stealing the message from the error here, and that's why it failed freeing below with a double free.

pbor replied:
It is *always* my fault :)
